tests/trainer/test_trainer.py

```diff
@@ -2146,3 +2146,26 @@ def test_dataloaders_are_not_loaded_if_disabled_through_limit_batches(running_st
         else getattr(trainer, f"{dl_prefix}_dataloaders")
     )
     assert dl is None
+
+
+@pytest.mark.parametrize(
+    ["trainer_kwargs", "expected_device_ids"],
+    [
+        ({}, [0]),
+        ({"devices": 1}, [0]),
+        ({"accelerator": "gpu", "devices": 1}, [0]),
```

Comment:
Please add tests for TPUs and IPUs, you could mock the availability

Reply:
The TPU have a lot of things to mock. It uses functions like `xm.xla_device` frequently not only we have to mock `_TPU_AVAILABLE`. I put more test cases in test_tpu.py and only run when we have tpus available. Should be similar for testing purposes. 

let me know what you think!

 https://github.com/PyTorchLightning/pytorch-lightning/blob/15364c18c81909e213f1910a57982f100f8a356d/pytorch_lightning/strategies/single_tpu.py#L44

Reply:
I agree! for TPU it might not be that trivial. You can test it on TPU directly.